File: Scala/src/main/scala/synapses/model/netElems/Layer.scala
package synapses.model.netElems

import synapses.model.netElems.Neuron.NeuronSerialized

object Layer {

  type Layer = LazyList[Neuron]

  def init(inputSize: Int,
           outputSize: Int,
           activation: Activation)
          (weightInitF: => Double)
  : Layer = LazyList
    .range(0, outputSize)
    .map { _ =>
      Neuron.init(
        inputSize,
        activation
      )(weightInitF)
    }

  def output(input: LazyList[Double])
            (layer: Layer)
  : LazyList[Double] =
    layer.map(Neuron.output(input))

  def backPropagated(learningRate: Double,
                     input: LazyList[Double],
                     outputWithErrors: LazyList[(Double, Double)])
                    (layer: Layer)
  : (LazyList[Double], Layer) = {
    val (errorsMulti, newNeurons) = outputWithErrors
      .zip(layer)
      .map { case (outWithError, neuron) =>
        Neuron.backPropagated(
          learningRate,
          input,
          outWithError
        )(neuron)
      }
      .unzip
    val inErrors =
      errorsMulti.foldLeft(LazyList.continually(0.0)) {
        case (acc, x) => acc
          .zip(x)
          .map(t => t._1 + t._2)
      }
    (inErrors, newNeurons)
  }

  type LayerSerialized = List[NeuronSerialized]

  def serialized(layer: Layer): LayerSerialized =
    layer
      .map(Neuron.serialized)
      .toList

  def deserialized(layerSerialized: LayerSerialized): Layer =
    layerSerialized
      .map(Neuron.deserialized)
      .to(LazyList)

}

File: implementation/Python/model/net_elems/layer.py
from itertools import count
from typing import Callable, Tuple, List
from functional import seq
from functional.pipeline import Sequence

from model import utilities
from model.net_elems import activation
from model.net_elems import neuron
from model.net_elems.neuron import NeuronSerialized

Layer = Sequence


def init(input_size: int,
         output_size: int,
         activation_f: activation.Activation,
         weight_init_f: Callable[[], Callable[[], float]]
         ) -> Layer:
    return seq \
        .range(output_size) \
        .map(lambda _:
             neuron.init(input_size,
                         activation_f,
                         weight_init_f()))


def output(input: Sequence, layer: Layer) -> Sequence:
    return layer.map(lambda x: neuron.output(input, x))


def back_propagated(learning_rate: float,
                    input: Sequence,
                    output_with_errors: Sequence,
                    layer: Layer
                    ) -> Tuple[Sequence, Layer]:
    errors_multi_with_new_neurons = output_with_errors \
        .zip(layer) \
        .map(lambda t: neuron.back_propagated(learning_rate, input, t[0], t[1]))
    (errors_multi, new_neurons) = utilities \
        .lazy_unzip(errors_multi_with_new_neurons)
    in_errors = errors_multi.fold_left(
        seq(count(0)).map(lambda _: 0.0),
        lambda acc, x: acc.zip(x).map(lambda t: t[0] + t[1])
    )
    return in_errors, new_neurons


LayerSerialized = List[NeuronSerialized]


def serialized(layer: Layer) -> LayerSerialized:
    return layer \
        .map(lambda x: neuron.serialized(x)) \
        .to_list()


def deserialized(layer_serialized: LayerSerialized) -> Layer:
    return seq(layer_serialized) \
        .map(lambda x: neuron.deserialized(x))
